syntax = "proto3";

package card.v1;

import "google/api/annotations.proto";
import "google/protobuf/empty.proto";
import "protoc-gen-openapiv2/options/annotations.proto";

option go_package = "github.com/klaital/cardsorter/backend/gen/protos;cardv1";

option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_swagger) = {
  info: {
    title: "Card Service API";
    version: "1.0";
    description: "API for managing cards in your collections";
    contact: {
      name: "Card Service Team";
      url: "https://github.com/klaital/cardsorter";
    };
  }
};


service CardService {
  // CreateCard creates a new card in a library
  rpc CreateCard(CreateCardRequest) returns (CreateCardResponse) {
    option (google.api.http) = {
      post: "/v1/libraries/{library_id}/cards"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "Create a new card"
      description: "Creates a new card in the specified library"
      tags: ["cards"]
    };
  }

  // GetCards retrieves all cards in a library
  rpc GetCards(GetCardsRequest) returns (GetCardsResponse) {
    option (google.api.http) = {
      get: "/v1/libraries/{library_id}/cards"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "Fetch the cards in the specified library"
      description: "Fetch the cards in the specified library"
      tags: ["cards"]
    };

  }

  // GetCard retrieves a specific card
  rpc GetCard(GetCardRequest) returns (GetCardResponse) {
    option (google.api.http) = {
      get: "/v1/libraries/{library_id}/cards/{card_id}"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "Fetch the details for a specific card from a library"
      description: "Fetch the details for a specific card from a library"
      tags: ["cards"]
    };
  }

  // MoveCard moves a card to a different library
  rpc MoveCard(MoveCardRequest) returns (google.protobuf.Empty) {
    option (google.api.http) = {
      post: "/v1/libraries/{library_id}/cards/{card_id}/move"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "Move a card from one library to another"
      description: "Move a card from one library to another"
      tags: ["cards"]
    };
  }

  // DeleteCard removes a card from a library
  rpc DeleteCard(DeleteCardRequest) returns (google.protobuf.Empty) {
    option (google.api.http) = {
      delete: "/v1/libraries/{library_id}/cards/{card_id}"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "Delete all copies of a card from a library."
      description: "Delete all copies of a card from a library."
      tags: ["cards"]
    };
  }
}

message Card {
  int64 id = 1;
  int64 library_id = 2;
  string name = 3;
  string set_name = 4 [(grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
    description: "Set ID, e.g. 'FIN'"
  }];
  string condition = 5 [(grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
    description: "Description of the card condition"
  }];
  bool foil = 6;
  string collector_number = 7 [(grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
    description: "The collection number for this card, e.g. '0210'"
  }];
  int32 usd_price = 8;
  int32 qty = 9;
}

message CreateCardRequest {
  int64 library_id = 1;
  string name = 2;
  string set_name = 3;
  string condition = 4;
  bool foil = 5;
  string collector_number = 6;
  int32 usd_price = 7;
}

message CreateCardResponse {
  Card card = 1;
}

message GetCardsRequest {
  int64 library_id = 1;
}

message GetCardsResponse {
  repeated Card cards = 1;
}

message GetCardRequest {
  int64 library_id = 1;
  int64 card_id = 2;
}

message GetCardResponse {
  Card card = 1;
}

message MoveCardRequest {
  int64 library_id = 1;
  int64 card_id = 2;
}

message DeleteCardRequest {
  int64 library_id = 1;
  int64 card_id = 2;
}
